kv: redirect on server-side RangeKeyMismatch when possible

kvoli opened the pull request:
Previously, `RangeKeyMismatchError` would construct a `RangeInfo`
without a `ClosedTimestampPolicy`. This caused `ClosedTimestampPolicy`
to default to `LAG_CLUSTER_BY_SETTING` when there may be a different
policy currently. To address this, this patch adds `ClosedTimestampPolicy`
as an argument for constructing `RangeKeyMismatchError`. This change is
extended to include the entire `RangeInfo`, containing `ClosedTimestampPolicy`
in `RangeKeyMismatchError.AppendRangeInfo()`.

Previously, `RangekeyMismatchError` would be returned when a
`BatchRequest` was directed at the wrong Range. This caused client side retries,
even if the replica existed on the store servicing the `BatchRequest`.
`store_send` will now check and retry locally when possible, instead of returning
a `RangekeyMismatchError`. Updated `RangeInfos` are now returned in the
`BatchResponse`, not modifying client `RangeDescriptiorCache` invalidation
behavior upon response. This change consolidates tests which assume server side
retries are not possible on kv requests.

resolves cockroachdb#72400

Release note (performance improvement): Stores will now server side
retry requests that are directed at the incorrect Range, however
still within the store. This commonly occurs following a Range split,
where the RHS has not yet been rebalanced away. In which case, the RHS
Range will be sitting adjacent to the LHS Range, on the same store.
Server side retries reduce KV hops, following a mismatch from 2 to 1.
This has the effect of reducing tail latency in the retriable case.
If the requested Range straddles both the LHS and RHS, no server
side retry is attempted.

kvoli committed:
kv: redirect on server-side RangeKeyMismatch

Previously, `RangekeyMismatchError` would be returned when a
`BatchRequest` was directed at the wrong Range.

This caused client side retries, even if the replica existed
on the store servicing the `BatchRequest`.

`store_send` will now check and retry locally when possible, instead of
returning a `RangekeyMismatchError`. Updated `RangeInfos` are now
returned in the `BatchResponse`, not modifying client
`RangeDescriptiorCache` invalidation behavior upon response.

This change consolidates tests which assume server side retries are not
possible on kv requests.

resolves #72400

Release note (performance improvement): Stores will now server side
retry requests that are directed at the incorrect Range, most commonly
following a recent range split. This patch has the effect of reducing
tail latency following Range splits.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75446)
<!-- Reviewable:end -->

kvoli committed:
kv: accept ClosedTimestampPolicy on RKey mismatch

Previously, `RangeKeyMismatchError` would construct a `RangeInfo`
without a `ClosedTimestampPolicy`. This caused `ClosedTimestampPolicy`
to default to `LAG_CLUSTER_BY_SETTING` when there may be a different
policy currently. To address this, this patch adds `ClosedTimestampPolicy`
as an argument for constructing `RangeKeyMismatchError`. This change is
extended to include the entire `RangeInfo`, containing `ClosedTimestampPolicy`
in `RangeKeyMismatchError.AppendRangeInfo()`.

Release note: None

nvanbenschoten commented on pkg/kv/kvserver/replica.go:
```diff
@@ -1489,8 +1494,7 @@ func (r *Replica) checkSpanInRangeRLocked(ctx context.Context, rspan roachpb.RSp
 		return nil
 	}
 	return roachpb.NewRangeKeyMismatchError(
-		ctx, rspan.Key.AsRawKey(), rspan.EndKey.AsRawKey(), desc, r.mu.state.Lease,
```
Drive-by comment — I intend to do a deeper review later: Isn't the replica already `RLocked` at this point? So won't this deadlock? Go's locks are not re-entrant, though this only result in a deadlock for re-entrant attempts on RLock if there was an intermediate call to Lock.

kvoli replied:
That seems right - thanks for pointing this out. It is trying to re-acquire a read lock in `GetClosedTimestampPolicy`. Updating.

nvb reviewed (commented):
> Reviewed 8 of 8 files at r1, 2 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75446)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @kvoli)

___
*[-- commits, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMfdxj5FAxuWdfLTEi):*
Could we expand on this release note to mention what this means for users? Something about reduced request tail latency during range splits.
___
*[pkg/kv/kvserver/replica.go, line 1497 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMbTPqCT0uIemLJXUy:-MuMbTPqCT0uIemLJXUz:b-z9i0y6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/replica.go#L1497)):*
> ```Go
> 	}
> 	return roachpb.NewRangeKeyMismatchError(
> 		ctx, rspan.Key.AsRawKey(), rspan.EndKey.AsRawKey(), desc, r.mu.state.Lease, r.ClosedTimestampPolicy())
> ```

Could we patch your fix to use `closedTimestampPolicyRLocked()` instead of `ClosedTimestampPolicy()` into the first commit where the bug was introduced, instead of the second?
___
*[pkg/kv/kvserver/replica_command.go, line 337 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMbiM_BlxI7YKHF_MR:-MuMbiM_BlxI7YKHF_MS:b-o2gtka) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/replica_command.go#L337)):*
> ```Go
> 			// correct range.
> 			if !kvserverbase.ContainsKey(desc, args.Key) {
> 				l, _ := r.GetLease()
> ```

Should we call `Replica.GetRangeInfo()` here instead of the two separate method calls which each acquire the mutex?
___
*[pkg/kv/kvserver/replica_protected_timestamp.go, line 190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMc1avDZGmVk7gXg4j:-MuMc1avDZGmVk7gXg4k:bl6fh0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/replica_protected_timestamp.go#L190)):*
> ```Go
> 	if !kvserverbase.ContainsKeyRange(desc, args.Key, args.EndKey) {
> 		return false, false, "", roachpb.NewRangeKeyMismatchError(ctx, args.Key, args.EndKey, desc,
> 			r.mu.state.Lease, r.ClosedTimestampPolicy())
> ```

We're holding the `RLock` here as well.
___
*[pkg/kv/kvserver/store_send.go, line 160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMef9_8geboNheXHdN:-MuMef9_8geboNheXHdO:b-l7kvql) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/store_send.go#L160)):*
> ```Go
> 	}
> 
> 	// Get range and add command to the range for execution.
> ```

Should this all be pulled into the loop as well? We'll want to update `ba.RangeID` on a retry and we'll also want the `IsInitialized` check, so it seems like this all belongs in the retry loop. That will also avoid the duplicate call to `repl.Send` and the need to replace `repl`.
___
*[pkg/kv/kvserver/store_send.go, line 195 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMcNAr4qbBrAnfKCV6:-MuMcNAr4qbBrAnfKCV7:bc17v1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/store_send.go#L195)):*
> ```Go
> 	}
> 
> 	// Force a retry loop on range key errors, if we are able to redirect the request
> ```

Does this new code belong in this commit?
___
*[pkg/kv/kvserver/store_send.go, line 196 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMcnMmAKUrrnqFYWTT:-MuMcnMmAKUrrnqFYWTU:b7dacb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/store_send.go#L196)):*
> ```Go
> 
> 	// Force a retry loop on range key errors, if we are able to redirect the request
> 	// to the correct range on this store on a RangeKeyMismatchError
> ```

Comments should use proper punctuation throughout. See https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences.
___
*[pkg/kv/kvserver/store_send.go, line 285 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMeye3EkBQoI2YM9vW:-MuMeye3EkBQoI2YM9vX:b99e0nd) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/store_send.go#L285)):*
> ```Go
> 			}
> 			// We have to write `t` back to `pErr` so that it picks up the changes.
> 			rangeInfos = t.Ranges
> ```

If we do redirect successfully, do we want to return all additional `rangeInfos`, including those that came from the lookahead, or just the one that originally came from the (possible multiple) `RangeKeyMismatchError`(s)? Will the approach we have here result in duplicate information about the same range?
___
*[pkg/roachpb/errors.go, line 563 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMc_AK3yJ5slt7KW0B:-MuMc_AK3yJ5slt7KW0C:bk4nq6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/roachpb/errors.go#L563)):*
> ```Go
> 	desc *RangeDescriptor,
> 	lease *Lease,
> 	ctspol RangeClosedTimestampPolicy,
> ```

nit: this is usually called `ctPolicy`.
___
*[pkg/kv/kvserver/helpers_test.go, line 484 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMcfyI6wvl3Q7UMO5e:-MuMcfyI6wvl3Q7UMO5f:b-oh7fg3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e27baa02ee9abc50d4445876e88c07181be77b55/pkg/kv/kvserver/helpers_test.go#L484)):*
> ```Go
> }
> 
> // ClosedTimestampPolicy returns the closed timestamp policy of the range, which
> ```

With the other proposed changes in this commit, I think we can revive this.


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75446)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @nvanbenschoten)

___
*[-- commits, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMfdxj5FAxuWdfLTEi):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could we expand on this release note to mention what this means for users? Something about reduced request tail latency during range splits.
</blockquote></details>

Updated to include note on reduces hops following range split - resulting in better tail latency perf.
___
*[pkg/kv/kvserver/replica.go, line 1497 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMbTPqCT0uIemLJXUy:-MuRAVK41yzb9POCtBnr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/replica.go#L1497)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could we patch your fix to use `closedTimestampPolicyRLocked()` instead of `ClosedTimestampPolicy()` into the first commit where the bug was introduced, instead of the second?
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/replica_command.go, line 337 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMbiM_BlxI7YKHF_MR:-MuS8ulv10Bp03tYLOMM:bcflsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/replica_command.go#L337)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we call `Replica.GetRangeInfo()` here instead of the two separate method calls which each acquire the mutex?
</blockquote></details>

updated to call `Replica.GetRangeInfo()`
___
*[pkg/kv/kvserver/replica_protected_timestamp.go, line 190 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMc1avDZGmVk7gXg4j:-MuS8xmC78o4FQ54Sgus:b-9t1pr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/replica_protected_timestamp.go#L190)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We're holding the `RLock` here as well.
</blockquote></details>

This was causing a possible deadlock - on re-entry with RLock promotion. Updated to use appropriate method.
___
*[pkg/kv/kvserver/store_send.go, line 160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMef9_8geboNheXHdN:-MuS99iv7h4LvpkwTwdU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/store_send.go#L160)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should this all be pulled into the loop as well? We'll want to update `ba.RangeID` on a retry and we'll also want the `IsInitialized` check, so it seems like this all belongs in the retry loop. That will also avoid the duplicate call to `repl.Send` and the need to replace `repl`.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/store_send.go, line 195 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMcNAr4qbBrAnfKCV6:-MuRB21h5JnJhIJRXyv-:bkhxqud) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/store_send.go#L195)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Does this new code belong in this commit?
</blockquote></details>

Updated to appropriate commit.
___
*[pkg/kv/kvserver/store_send.go, line 196 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMcnMmAKUrrnqFYWTT:-MuS9EK33nQfeXthLNWG:bqikcwc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/store_send.go#L196)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Comments should use proper punctuation throughout. See https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences.
</blockquote></details>

Updated to use correct comment format.
___
*[pkg/kv/kvserver/store_send.go, line 285 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMeye3EkBQoI2YM9vW:-MuS9J1y7rrAzmw4LBAz:be4bwli) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/store_send.go#L285)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If we do redirect successfully, do we want to return all additional `rangeInfos`, including those that came from the lookahead, or just the one that originally came from the (possible multiple) `RangeKeyMismatchError`(s)? Will the approach we have here result in duplicate information about the same range?
</blockquote></details>

In this case, I believe we want to return only from the RangeKeyMismatchErrors. The current approach always results in duplicate information from the replica, which I updated the TODO with.

An option here is to also update the `ClientRangeInfo` header within the `BatchRequest`, which is then used by the replica to decide whether to include `BatchResponse.RangeInfo` on a successful request.

Currently it should not be the case that `BatchResponse.RangeInfo`  returned from the `Replica` is not contained within `candidateRanges`, collected during `RangeKeyMismatchError`(s). However, previously - it was possible for this to occur due to `ClosedTimestampPolicy` not being included in the `RangeKeyMismatchError` constructor or append function.

I'm unsure how to proceed here - we could add a check to assert that `BatchResponse.RangeInfos` is a proper subset of `RangeKeyMismatchError.RangeInfos` or update `BatchRequest.ClientRangeInfo` on a retry and avoid duplicate `RangeInfos`.
___
*[pkg/roachpb/errors.go, line 563 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMc_AK3yJ5slt7KW0B:-MuSAlS7Bx5qm_pxJJkd:b-cypxrc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/roachpb/errors.go#L563)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: this is usually called `ctPolicy`.
</blockquote></details>

updated.
___
*[pkg/kv/kvserver/helpers_test.go, line 484 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMcfyI6wvl3Q7UMO5e:-MuSAockDvti4LrqKExm:bxhi25x) ([raw file](https://github.com/cockroachdb/cockroach/blob/e27baa02ee9abc50d4445876e88c07181be77b55/pkg/kv/kvserver/helpers_test.go#L484)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

With the other proposed changes in this commit, I think we can revive this.
</blockquote></details>

Added back in.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 10 of 10 files at r3, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75446)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @kvoli)

___
*[pkg/kv/kvserver/store_send.go, line 285 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMeye3EkBQoI2YM9vW:-MuSG6TJD6IIVWUJccqx:b-wljvhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/store_send.go#L285)):*
> An option here is to also update the ClientRangeInfo header within the BatchRequest, which is then used by the replica to decide whether to include BatchResponse.RangeInfo on a successful request.

This seems like a good idea.

If we did this, could we then omit the target `RangeInfo` from `candidateRanges` so that we can blindly `br.RangeInfos = append(candidateRanges, br.RangeInfos...)` on the response path without risking duplication except in the case of a race with a range merge?

I'm also wondering whether it's even a problem to have overlapping range infos in `br.RangeInfos`. It looks like the RangeCache on the client will just be updated with them one-by-one. So we'd just need to make sure that in the case of contradictions, the newer information comes last.
___
*[pkg/kv/kvserver/store_send.go, line 165 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuSF26jDWC7pKKdNJMk:-MuSF26jDWC7pKKdNJMl:b-y406so) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca35bbd5df809ae92b5748ccf5a4b21d0a0b078/pkg/kv/kvserver/store_send.go#L165)):*
> ```Go
> 	// Run a retry loop on retriable RangeKeyMismatchErrors, where the requested RangeID is
> 	// no longer correct, however we have the Range locally and can redirect the request.
> 	for retries, isRetriableRangeKeyMismatch := 0, true; retries < 2 && isRetriableRangeKeyMismatch; retries++ {
> ```

Why do we limit the number of retries to 2?
___
*[pkg/kv/kvserver/replica_test.go, line 879 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuSHVz6E4nsLZypNd6K:-MuSHVz6E4nsLZypNd6L:besy4uw) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca35bbd5df809ae92b5748ccf5a4b21d0a0b078/pkg/kv/kvserver/replica_test.go#L879)):*
> ```Go
> 
> 	// Expect server side retry on RangeKeyMismatch, which gets swallowed
> 	// at store_send.go:send
> ```

nit: punctuation here and in `store_test.go`


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75446)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @nvanbenschoten)

___
*[pkg/kv/kvserver/store_send.go, line 285 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuMeye3EkBQoI2YM9vW:-MuSa6lp4FUw84SdmV8j:b-itixwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffd1087e2f78cf54456f5497411e4bbd4212a08/pkg/kv/kvserver/store_send.go#L285)):*
> could we then omit the target RangeInfo from candidateRanges so that we can blindly br.RangeInfos = append(candidateRanges, br.RangeInfos...) on the response path without risking duplication except in the case of a race with a range merge?

I don't think we are able to omit the target `RangeInfo` from `candidateRanges`, unless `len(br.RangeInfo) > 0`. However with the suggested change, to update `ba.ClientRangeInfo`, it should no longer be the case that the replica will populate `ba.RangeInfos` - as it sees an updated `ClientRangeInfo` that is correct (unless there is a race). 

> I'm also wondering whether it's even a problem to have overlapping range infos in br.RangeInfos.

This was the initial approach, however it led to a test failing. With the update to include all `RangeInfo` fields in `RangekeyMismatchError`(s), it should no longer be possible (afaik) to have conflicting `RangeInfos` between the `candidateRanges` and `br.RangeInfos`.

> in the case of contradictions, the newer information comes last.

~~Is the`RangeCache` commutative? The case where we have identical RangeInfos in terms of time and generation ordering, yet differing `CLOSED_TIMESTAMP_POLICY` or another field seems like an issue that would affect ordering. This was previously occurring, prior to including all fields as mentioned above (despite LILO ordering, however I need to double check).~~ I can't gleam anything about this possibly occurring still, however you might have a better intuition.

I've included the suggested changes here and updated the comment. 
___
*[pkg/kv/kvserver/store_send.go, line 165 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuSF26jDWC7pKKdNJMk:-MuSQCHSD95OzpwMXx6V:bahk9ot) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca35bbd5df809ae92b5748ccf5a4b21d0a0b078/pkg/kv/kvserver/store_send.go#L165)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why do we limit the number of retries to 2?
</blockquote></details>

Unsure, this is an implicit assumption that perhaps should be explicit in the comment - after two RangeKeyMismatchError, if it were still retriable that would indicate we are chasing after recurrent splits.

Is there any reason to have an arbitrarily limited number of retries, I'm unsure on this one.
___
*[pkg/kv/kvserver/replica_test.go, line 879 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuSHVz6E4nsLZypNd6K:-MuSe-Pz19sQdMht-Fxg:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca35bbd5df809ae92b5748ccf5a4b21d0a0b078/pkg/kv/kvserver/replica_test.go#L879)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: punctuation here and in `store_test.go`
</blockquote></details>

Fixed.


<!-- Sent from Reviewable.io -->

irfansharif commented on pkg/kv/kvclient/kvcoord/dist_sender_test.go:
```diff
@@ -1370,8 +1370,8 @@ func TestEvictCacheOnError(t *testing.T) {
 	rangeMismachErr := roachpb.NewRangeKeyMismatchError(
-		context.Background(), nil, nil, &lhs, nil /* lease */)
-	rangeMismachErr.AppendRangeInfo(context.Background(), rhs, roachpb.Lease{})
+		context.Background(), nil, nil, &lhs, nil /* lease */, roachpb.LAG_BY_CLUSTER_SETTING)
```
[nit] Should we change the constructor here to take in a pointer instead? That way the caller doesn't need to bother with the verbose default. Alternatively, have another constructor for NewRangeKeyMismatchError that takes in an explicit closed-ts policy.

kvoli replied:
Updated to a new constructor, one including the closed timestamp policy and another not.

irfansharif commented on pkg/kv/kvserver/store_send.go:
```diff
@@ -157,104 +162,145 @@ func (s *Store) Send(
+	// RangeKeyMismatchErrors shouldn't be possible unless repeated splitting upon the range containing
+	// the request key span occurs. This seems unlikely, however as this behavior is degenerate retries
+	// is limited to at most 3 before returning an error to the client. Establish an appropriate bounds.
+	for retries, isRetriableRangeKeyMismatch := 0, true; retries < maxRKeyMismatchRetries && isRetriableRangeKeyMismatch; retries++ {
```
Checkout pkg/util/retry -- I'm curious if we can use that library here instead, maybe getting rid of maxRKeyMismatchRetries. Seeing as how this retry block is useful only in the presence of recurrent splits, which if it were to occur would induce retries at the client anything (but with additional RTTs), it'd be fine IMO to retry until the surrounding ctx gets cancelled.

irfansharif commented on pkg/kv/kvserver/store_send.go:
```diff
@@ -157,104 +162,145 @@ func (s *Store) Send(
+	// (1): On a RangeKeyMismatchError that is retriable.
+	// (2): On a successful batch request, where suggested ranges are returned
+	//      by the replica to update the client with. This is appended before returning.
+	var candidateRanges []roachpb.RangeInfo
```
From looking at this live: lets make sure that the loop body below cares only about populating this slice correctly. Only at the very end when returning either a batch-response or a pErr, be sure to fill the proto in with the accumulated range infos. This would help avoid "bugs" where we drop range infos unwittingly (which I think we're doing with our handling of pErr).

irfansharif commented on pkg/kv/kvserver/store_send.go:
```diff
@@ -157,104 +162,145 @@ func (s *Store) Send(
+			// to the client, to invalidate their cache. When the batch request has a
+			// mismatch and is not retriable, this information is instead returned
+			// inside pErr.RangeInfos.
+			if retries > 0 {
```
Do we need this conditional? If there are not retries, I think our set of candidate range infos would be empty. This relates to the point above -- we could simplify this a bit if all the retry block did was accumulate the set of range infos gathered across all the retries, and only at the very end populate the right proto (BatchResponse or PErr).

kvoli replied:
It is redundant. Updated to accumulate - the only thing that is a bit iffy here, for the approach of populating pErr at the end is the type of the error needs to be known. So doing this at the very end wouldn't make sense as it would require another switch. I've still updated it to be more explicit, however not all at the end.

irfansharif commented on pkg/kv/kvserver/store_send.go:
```diff
@@ -157,104 +162,145 @@ func (s *Store) Send(
+					// duplicate BatchResponse.RangeInfos upon a successful retry for the client
+					// to invalidate their RangeCache with.
+					isRetriableRangeKeyMismatch = true
+					ba.RangeID = ri.Desc.RangeID
```
[nit] I think the code here is correct, but it's not obvious at first glance given we're mutating the same request proto over multiple iterations. It would be a bit easier I think if we maintained a fresh copy of the request for each retry, populating it correctly based on the information we had most recently.

kvoli replied:
Good idea, done.

irfansharif reviewed (commented):
Nice work! First commit LGTM, left a few minor suggestions to improve readability for the second. I'd also suggest dumbing down the release note slightly so it's understandable to end users that aren't familiar with the internal mechanics in place (RHS, LHS replicas, range splits, rebalances). It'd be clearer I think to talk in terms of the end user impact (improved tail, like you correctly point out), maybe with the fewest words for how we got there.

irfansharif commented on pkg/kv/kvserver/store_send.go:
```diff
@@ -157,104 +158,166 @@ func (s *Store) Send(
+
+			// Update the RangeInfos that will be returned to the client, to invalidate
+			// the client's RangeCache.
+			t.AppendRangeInfo(ctx, candidateRanges...)
```
Between this and the line above where we populate candidateRanges using t.Ranges:

```go
candidateRanges = append(candidateRanges, t.Ranges...)
```

Are we including duplicates? Do we instead want to write to t.Ranges directly?

irfansharif commented on pkg/kv/kvserver/store_send.go:
```diff
@@ -157,104 +158,166 @@ func (s *Store) Send(
+				// if returned from the replica. Note here that newer ranges are always appended, so that the
+				// oldest RangeInfo is processed by the Client's RangeCache first, which is then invalidated on
+				// conflict by newer data (LILO).
+				ri.ValidateRangeInfoLease(ctx)
```
Out of curiosity, what happens if we push this validation code into (*Replica).GetRangeInfo? Seems like a sharp edge to have to validate what's returned when it could be done at the point of return itself.

kvoli replied:
Seems like this validation code already exists in that method. I've updated to remove here and with the separation.

irfansharif commented on pkg/kv/kvserver/store_send.go:
```diff
@@ -157,104 +158,166 @@ func (s *Store) Send(
+			}
+
+			// if we are able to retry, iterate retry loop.
+			if isRetriableRangeKeyMismatch {
```
At the cost of adding another loop over `ris`, we could do the following which I think would be easier to read + avoid the need for `isRetriableRangeKeyMismatch`.

```go
for _, ri := range ris {
  candidateRanges = append(candidateRanges, ri)
}

for _, ri := range ris {
  if ri.Desc.RSpan().ContainsKeyRange(rSpan.Key, rSpan.EndKey) {
    baRetry = ba // ... and other stuff
    candidateRanges = append(candidateRanges, t.Ranges...) // collect for next loop iteration, instead of where we're doing it now
    continue
  }
}
```

In fact, the first loop is actually a one liner.
```go
candidateRanges = append(candidateRanges, ris...)
```

[nit] I'd also suggest s/candidateRanges/rangeInfos for two reasons:
- They're not ranges, which are a distinct separate thing
- They're not candidates, we're going to unconditionally return everything in the slice, whether through the batch response or through pErr.

kvoli replied:
I like the idea of splitting. The only issue I have is I don't think it is possible to nested continue here. Instead, it will just continue the inner loop. Otherwise, I'll add.

irfansharif commented on pkg/kv/kvserver/replica_test.go:
```diff
@@ -871,16 +885,26 @@ func TestReplicaRangeBoundsChecking(t *testing.T) {
 		RangeID: 1,
-	}, &gArgs)
+	}
+	ba.AdmissionHeader = roachpb.AdmissionHeader{}
```
Is this necessary?

kvoli replied:
Removed.

irfansharif commented on pkg/kv/kvserver/replica_test.go:
```diff
@@ -871,16 +885,26 @@ func TestReplicaRangeBoundsChecking(t *testing.T) {
+	br, pErr := tc.store.Send(context.Background(), ba)
+
+	if pErr != nil {
+		t.Errorf("Unexpected range key mismatch error: %s", pErr)
```
If all we're doing is checking whether pErr is non-nil, it's not necessary that this is a range key mismatch error, right? It sounds like we want `require.NotNil(t, pErr)` here instead.

kvoli replied:
We want it to be nil here. Updated to `assertNil`.

kvoli commented on pkg/kv/kvserver/store_send.go:
```diff
@@ -157,104 +158,166 @@ func (s *Store) Send(
+
+	// Avoid mutating the original request, instead creating an immutable BatchRequest copy on
+	// each retry.
+	baRetry := ba
```
Done.

kvoli commented on pkg/kv/kvserver/store_send.go:
```diff
@@ -157,104 +158,166 @@ func (s *Store) Send(
+	// not match any Range. The reuqest is retriable when the correct Range, corresponding to the
+	// request key span also exists entirely on this store, on a different Replica then requested.
+	retryOptions := retry.Options{
+		InitialBackoff: time.Nanosecond,
```
Yeah not too sure on sane defaults here. I don't think the default is suitable however? 

```go
opts.InitialBackoff = 50 * time.Millisecond             
opts.MaxBackoff = 2 * time.Second                             
opts.RandomizationFactor = 0.15                                                    
opts.Multiplier = 2                                                                                    
```

nvb reviewed (commented):
> Reviewed 5 of 10 files at r8, 1 of 5 files at r10, 3 of 4 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75446)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @kvoli, and @nvanbenschoten)

___
*[pkg/kv/kvserver/store_send.go, line 165 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuSF26jDWC7pKKdNJMk:-MuqxFp26rmw2toiA2Sp:b3kl7hr) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca35bbd5df809ae92b5748ccf5a4b21d0a0b078/pkg/kv/kvserver/store_send.go#L165)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Unsure, this is an implicit assumption that perhaps should be explicit in the comment - after two RangeKeyMismatchError, if it were still retriable that would indicate we are chasing after recurrent splits.

Is there any reason to have an arbitrarily limited number of retries, I'm unsure on this one.
</blockquote></details>

I don't see a reason to limit the number of retries.
___
*[pkg/kv/kvserver/store_send.go, line 170 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuBxQx7-4cyStsAt5Y5-r9-170:-Muqx2947WLLrmBGzxsw:b151bgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f04a177cce0a2772ca073774327c388f822ebe34/pkg/kv/kvserver/store_send.go#L170)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Why do we need to copy `ba`? Was this causing issues? `ba` is a value that should not escape the stack of this function.
___
*[pkg/kv/kvserver/store_send.go, line 176 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuBxQx7-4cyStsAt5Y5-r9-176:-Muqwp0wCJGwIWhT2ERB:b973pjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f04a177cce0a2772ca073774327c388f822ebe34/pkg/kv/kvserver/store_send.go#L176)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

I'm not too certain on sane numbers here, the defaults however appear too large for local retries; better intended for network hops.

```go
opts.InitialBackoff = 50 * time.Millisecond             
opts.MaxBackoff = 2 * time.Second                             
opts.RandomizationFactor = 0.15                                                    
opts.Multiplier = 2                                                                                    
```

I've tentatively updated to a uniform 

```go
opts.InitialBackoff = 100 * time.Nanoseconds // (0.1ms)         
opts.MaxBackoff = 100 * time.Nanoseconds // (0.1ms)                
```
</blockquote></details>

Why do we need a backoff for this retry? When a range splits, its RHS should be created atomically on the same store. Were we seeing thrashing?
___
*[pkg/kv/kvserver/store_send.go, line 215 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuqxSRzEuqYDFszLuim:-MuqxSRzEuqYDFszLuin:b-9xgf3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/839861d0df4535ceb34ee449dbac6f0ad0105077/pkg/kv/kvserver/store_send.go#L215)):*
> ```Go
> 			// and pass these to the client, to invalidate their cache. This is
> 			// otherwise empty.
> 			br.RangeInfos = append(rangeInfos, br.RangeInfos...)
> ```

Let's only do this if `rangeInfos` is not empty, otherwise this will lead to wasted heap allocations.

Also, explain why we're appending in this order instead of `br.RangeInfos = append(br.RangeInfos, rangeInfos...)`, which wouldn't need the empty check to avoid a heap allocation.
___
*[pkg/kv/kvserver/store_send.go, line 289 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-Muqy07c1tvKhEYZowD0:-Muqy07c1tvKhEYZowD1:b6icr4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/839861d0df4535ceb34ee449dbac6f0ad0105077/pkg/kv/kvserver/store_send.go#L289)):*
> ```Go
> 					baCopy = ba
> 					baCopy.RangeID = ri.Desc.RangeID
> 					baCopy.ClientRangeInfo.ClosedTimestampPolicy = ri.ClosedTimestampPolicy
> ```

nit: `ba.ClientRangeInfo = roachpb.ClientRangeInfo{ ... }` makes it more clear that the intention is to update the entire struct.  
___
*[pkg/kv/kvserver/store_send.go, line 306 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuqyJKhE0SorreHdaUn:-MuqyJKhE0SorreHdaUo:b-2lrzvc) ([raw file](https://github.com/cockroachdb/cockroach/blob/839861d0df4535ceb34ee449dbac6f0ad0105077/pkg/kv/kvserver/store_send.go#L306)):*
> ```Go
> 			// to invalidate their RangeCache and update the request headers. Include all
> 			// RangeInfos that have been accumulated so far.
> 			t.AppendRangeInfo(ctx, rangeInfos...)
> ```

Is this appending potentially stale rangeInfos that could overlap with the more up-to-date information in `t`?


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75446)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @kvoli, and @nvanbenschoten)

___
*[pkg/kv/kvserver/store_send.go, line 176 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuBxQx7-4cyStsAt5Y5-r9-176:-Mur7gqeBuT9Ar5MCWoU:b-53emys) ([raw file](https://github.com/cockroachdb/cockroach/blob/f04a177cce0a2772ca073774327c388f822ebe34/pkg/kv/kvserver/store_send.go#L176)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why do we need a backoff for this retry? When a range splits, its RHS should be created atomically on the same store. Were we seeing thrashing?
</blockquote></details>

This was a suggestion to cleanup the logic. There wasn't any particular reason other than cleanliness. I didn't see thrashing here.
___
*[pkg/kv/kvserver/store_send.go, line 215 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuqxSRzEuqYDFszLuim:-Mur7eMXDw2w3yUKZyPV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/839861d0df4535ceb34ee449dbac6f0ad0105077/pkg/kv/kvserver/store_send.go#L215)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Let's only do this if `rangeInfos` is not empty, otherwise this will lead to wasted heap allocations.

Also, explain why we're appending in this order instead of `br.RangeInfos = append(br.RangeInfos, rangeInfos...)`, which wouldn't need the empty check to avoid a heap allocation.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/store_send.go, line 289 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-Muqy07c1tvKhEYZowD0:-Mur7O51EhnJXuADiiee:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/839861d0df4535ceb34ee449dbac6f0ad0105077/pkg/kv/kvserver/store_send.go#L289)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: `ba.ClientRangeInfo = roachpb.ClientRangeInfo{ ... }` makes it more clear that the intention is to update the entire struct.  
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/store_send.go, line 306 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuqyJKhE0SorreHdaUn:-MurFb1H58YTudgfE_Hg:b-vxnihv) ([raw file](https://github.com/cockroachdb/cockroach/blob/839861d0df4535ceb34ee449dbac6f0ad0105077/pkg/kv/kvserver/store_send.go#L306)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this appending potentially stale rangeInfos that could overlap with the more up-to-date information in `t`?
</blockquote></details>

This is correct. Updated to prepend.


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75446)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @kvoli, and @nvanbenschoten)

___
*[pkg/kv/kvserver/store_send.go, line 170 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuBxQx7-4cyStsAt5Y5-r9-170:-MurGRI-15--znOnT3SC:bsay8mm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f04a177cce0a2772ca073774327c388f822ebe34/pkg/kv/kvserver/store_send.go#L170)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why do we need to copy `ba`? Was this causing issues? `ba` is a value that should not escape the stack of this function.
</blockquote></details>

This was done for tracing purposes, however I've updated to avoid any confusion.


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75446)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @kvoli, and @nvanbenschoten)

___
*[pkg/kv/kvserver/store_send.go, line 165 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuSF26jDWC7pKKdNJMk:-MurGZAOCcuFzdPXQawT:bc8fezp) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca35bbd5df809ae92b5748ccf5a4b21d0a0b078/pkg/kv/kvserver/store_send.go#L165)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I don't see a reason to limit the number of retries.
</blockquote></details>

In that case, would you suggest just looping and erroring (implicitly) on context cancellation?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75446)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @kvoli, and @nvanbenschoten)

___
*[pkg/kv/kvserver/store_send.go, line 165 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuSF26jDWC7pKKdNJMk:-MurHYLi62GEWHU54nC_:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/aca35bbd5df809ae92b5748ccf5a4b21d0a0b078/pkg/kv/kvserver/store_send.go#L165)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

In that case, would you suggest just looping and erroring (implicitly) on context cancellation?
</blockquote></details>

:thumbsup: 
___
*[pkg/kv/kvserver/store_send.go, line 176 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuBxQx7-4cyStsAt5Y5-r9-176:-MurHH653ht23pABB-bo:bnf5br5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f04a177cce0a2772ca073774327c388f822ebe34/pkg/kv/kvserver/store_send.go#L176)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

This was a suggestion to cleanup the logic. There wasn't any particular reason other than cleanliness. I didn't see thrashing here.
</blockquote></details>

If we don't need to or want to backoff then we probably shouldn't use the `retry` package. A `for { ... }` loop should be sufficient. 


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75446)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @nvanbenschoten)

___
*[pkg/kv/kvserver/store_send.go, line 170 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuBxQx7-4cyStsAt5Y5-r7-170:-MurIa2l7IC8IPD-l6UA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d619b5836ebbf7fbb3f3964378b7d7764095ff6/pkg/kv/kvserver/store_send.go#L170)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

From looking at this live: lets make sure that the loop body below cares only about populating this slice correctly. Only at the very end when returning either a batch-response or a pErr, be sure to fill the proto in with the accumulated range infos. This would help avoid "bugs" where we drop range infos unwittingly (which I think we're doing with our handling of pErr).
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/store_send.go, line 178 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuBxQx7-4cyStsAt5Y5-r7-178:-MurIcKkDdY_vJpC0oRr:bwy893x) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d619b5836ebbf7fbb3f3964378b7d7764095ff6/pkg/kv/kvserver/store_send.go#L178)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Checkout pkg/util/retry -- I'm curious if we can use that library here instead, maybe getting rid of maxRKeyMismatchRetries. Seeing as how this retry block is useful only in the presence of recurrent splits, which if it were to occur would induce retries at the client anything (but with additional RTTs), it'd be fine IMO to retry until the surrounding ctx gets cancelled.
</blockquote></details>

Following the other thread w/ @nvanbenschoten. Opting to use `for {...}` here as it doesn't seem time bound (beyond ctx cancellation) and we don't need the retry capabilities.
___
*[pkg/kv/kvserver/store_send.go, line 176 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/75446#-MuBxQx7-4cyStsAt5Y5-r9-176:-MurIPIK6KLXiHj5GyAP:baakbhx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f04a177cce0a2772ca073774327c388f822ebe34/pkg/kv/kvserver/store_send.go#L176)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If we don't need to or want to backoff then we probably shouldn't use the `retry` package. A `for { ... }` loop should be sufficient. 
</blockquote></details>

Updated to use `for {...}`.


<!-- Sent from Reviewable.io -->

irfansharif commented on pkg/roachpb/errors.go:
```diff
@@ -579,11 +576,37 @@ func NewRangeKeyMismatchError(
+		end,
+		desc,
+		lease,
+		LAG_BY_CLUSTER_SETTING, /* defaut closed timestsamp policy*/
```
[nit] Missing a space: "policy*/" + typo for "defaut"

irfansharif commented on pkg/kv/kvserver/replica_test.go:
```diff
@@ -854,6 +854,20 @@ func TestLeaseReplicaNotInDesc(t *testing.T) {
 }
 
+// TestReplicaRangeBoundsChecking tests two behaviors that should occur.
+// (1): Following a Range split, however prior to the RHS Range being
```
This is really helpful detail! And a good test for what we've added. I'd suggest renaming it a bit more appropriately given what it now does. Also [nit] there's something off with the first two sentences here. How about the following?

```
- Following a Range split, the client may send BatchRequests based on stale
  cache data targeting the wrong range. Internally this triggers a
  RangeKeyMismatchError, but in the cases where the RHS of the range is still
  present on the local store, we opportunistically retry server-side by
  re-routing the request to the right range. No error is bubbled up to the
  client.
- This test also ensures that after a successful server-side retry attempt we
  bubble up the most up-to-date RangeInfos for the client to update its range
  cache.
```

irfansharif commented on pkg/kv/kvserver/client_split_test.go:
```diff
@@ -496,11 +496,16 @@ func TestStoreRangeSplitAtRangeBounds(t *testing.T) {
-	if _, ok := pErr.GetDetail().(*roachpb.RangeKeyMismatchError); !ok {
-		t.Fatalf("expected RangeKeyMismatchError, found: %v", pErr)
+	// An AdminSplit request sent to the end of the old range should be re-routed
+	// to the start of the new range, succeeding but no new ranges should be created
```
s/but no ../but without creating any new ranges, also missing period at the end.

kvoli replied:
Updated.

irfansharif commented on pkg/kv/kvserver/store_send.go:
```diff
@@ -157,104 +157,162 @@ func (s *Store) Send(
+	var rangeInfos []roachpb.RangeInfo
+
+	// Run a retry loop on retriable RangeKeyMismatchErrors, where the requested RangeID does
+	// not match any Range. The reuqest is retriable when the correct Range, corresponding to the
```
[nit] Typo: "reuqest". There's something slightly off with this second sentence.

kvoli replied:
Fixed, updated second sentence to be parsed more easily.

irfansharif reviewed (approved):
LGTM! Nice work dude.
